File: code/cpp/1248.cpp
# https://leetcode.com/problems/count-number-of-nice-subarrays/
class Solution {
public:
    int numberOfSubarrays(vector<int>& nums, int k) {
        const int n = nums.size();
        nums.emplace_back(1);
        vector<int> count(n + 1, 1);
        int res = 0, countOdd = 0, countOddW = 0;
        for (int l = 0; l <= n; l++) {
            if (nums[l] & 1) {
                countOdd++;
                if (countOddW == k)
                    res += count[countOdd - k - 1] * count[countOdd - 1];
                else
                    countOddW++;
            } else
                count[countOdd]++;
        }
        return res;
    }
};

File: code/python/1248.py
# https://leetcode.com/problems/count-number-of-nice-subarrays/
class Solution:
    def numberOfSubarrays(self, nums: List[int], k: int) -> int:
        n = len(nums)
        nums.append(1)
        count = [1] * (n + 1)
        res = countOdd = countOddW = 0
        for l in range(n + 1):
            if nums[l] & 1:
                countOdd += 1
                if countOddW == k:
                    res += count[countOdd - k - 1] * count[countOdd - 1]
                else:
                    countOddW += 1
            else:
                count[countOdd] += 1
        return res
